### ISSUE TITLE
Corrupt PNG output from Windows build

### ISSUE THREAD

**Issue by Adreitz** [fhanau/Efficient-Compression-Tool#1]
I tried this tool on two images, a small palettized PNG and a large full-color PNG. Both had already been pretty well-compressed with other tools. As this was my first time using this optimizer, I made copies of the files and then stepped through the various settings to see what would happen. I used --strict -M# on both files.

Starting with the small file, -M1 made no changes, as I expected. -M2, however, unexpectedly produced a large reduction of almost 42% (61.5 KB removed). -M3 through -M5 then made no further changes, all saying "ECT: error: bad adaptive filter value". In addition, -M5 reported, "Decoding error 58: invalid ADLER32 encountered (checking ADLER32 can be disabled)". I opened the file in an image viewer to discover that the lines of the image looked like they had been scrambled, while the colors and large-scale order mostly seemed to be intact.

I tried again with a fresh copy of the small image, going immediately to --strict -M5, which produced exactly the same results. With another fresh copy, I tried removing the --strict flag, but this again produced the same results.

Thinking that perhaps ECT had a bug with its support for palettized images, I tried again with a larger, full-color PNG. Again, --strict -M1 made no changes and -M2 produced a large reduction of 31.6% (468 KB). Opening the file immediately, I saw that it was mostly black with bright R/G/B/C/M/Y/W pixels corresponding to edges of objects and gradients in the original image.

So, rather than failing on a certain type of PNG, it looks like there is some more general failure with reading or writing proper PNG files. Since this immediately corrupts the source image with no notification, I'd suggest trying to fix this quickly.

See attached for the images I was working with.

Thanks,
Aaron

Windows 7 SP1

![small original](https://cloud.githubusercontent.com/assets/11577741/9549060/0e87cd46-4d72-11e5-8a0c-01d10fb84c1c.png)
![small corrupt](https://cloud.githubusercontent.com/assets/11577741/9549062/0e8bdae4-4d72-11e5-84b1-4a329e2f75c2.png)
![large original](https://cloud.githubusercontent.com/assets/11577741/9549059/0e87b9aa-4d72-11e5-99ef-47c2d7f3a445.png)
![large corrupt](https://cloud.githubusercontent.com/assets/11577741/9549063/0e8cf50a-4d72-11e5-8eaf-64ec47979732.png)

**Comment by fhanau** [comment]
Thanks for the report.

I partially fixed the bug in the latest version. The problem was that a function wasn't 32-bit compatible. Unfortunately, almost all testing was done using 64-bit Mac OS X, so I couldn't find it earlier.
With the latest the latest comment, Image one stays the same. Image 2 SHOULD stay the same, but I didn't test that extensively.

By the way, which optimizer did you use for the images? Although ECT was intended to be quick at a high compression level, and not to achieve maximum compression, it should be difficult to compress better than -M5.

**Comment by Adreitz** [comment]
Thanks, Felix. I've tried the updated Windows build and didn't have a problem with these two files. I'll let you know if I see any problem with other files.

Regarding the programs I use, I actually have a fairly complex procedure to squeeze images like this (especially for palettized images). I can break it down for you later if you're interested, but the list of programs is:

PNGPalTest (for palettized images only)
Pixelformer (for palettized images only)
TruePNG
PNGOUT
pngwolf
huffmix
rezop
zopflipng (Mr_KrzYch00's mod from encode.ru)
deflopt

The true color image I attached is only lightly compressed (I only used TruePNG and PNGOUT), but I am currently compressing the palettized image using my full procedure and am down to 149951 bytes.

Thanks,
Aaron
